tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,231 @@
+
+            return foundFields.ToArray();
+        }
+
+        private static string SanitizeAsyncHoistedLocalName(string fieldName, int indexOfStateMachineSuffix)
+        {
+#if NETCOREAPP3_0_OR_GREATER
+            return fieldName.AsSpan()[1..indexOfStateMachineSuffix].ToString();
+#else
+            return fieldName.Substring(1, indexOfStateMachineSuffix - 1);
+#endif
+        }
+
+        private static bool IsStateMachineField(FieldInfo field)
+        {
+            if (field.Name.StartsWith(StateMachineFieldsPrefix))
+            {
+                AsyncMethodDebuggerInvoker.Log.Information($"Ignoring local (state machine field):  {field.FieldType.Name} {field.DeclaringType?.Name}.{field.Name}");
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsKnownStateMachineField(FieldInfo field)
+        {
+            // builder (various object types), state (int) and "this"
+            if (field.Name is BuilderFieldName or StateFieldName or StateMachineThisFieldName or StateMachineLiveDebuggerAddedField)
+            {
+                return true;
+            }
+
+            // awaiter etc. (e.g. <>u__1)
+            if (field.Name.StartsWith(StateMachineFieldsPrefix))
+            {
+                if (field.FieldType.Name is "TaskAwaiter" or "TaskAwaiter`1" or "ValueTaskAwaiter" or "ValueTaskAwaiter`1")
+                {
+                    return true;
+                }
+                else
+                {
+                    AsyncMethodDebuggerInvoker.Log.Information($"Ignoring local {field.FieldType.Name} {field.DeclaringType?.Name}.{field.Name}");
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static MethodBase GetAsyncKickoffMethod(Type stateMachineType)
+        {
+            var originalType = stateMachineType.DeclaringType;
+            var allMethods = originalType?.GetMethods(BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy);
+            MethodBase foundMethod = null;
+            for (int i = 0; i < allMethods?.Length; i++)
+            {
+                var currentMethod = allMethods[i];
+                if (currentMethod.GetCustomAttribute<AsyncStateMachineAttribute>()?.StateMachineType == stateMachineType)
+                {
+                    foundMethod = currentMethod;
+                    break;
+                }
+            }
+
+            return foundMethod;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static object GetAsyncKickoffThisObject<TTarget>(TTarget stateMachine)
+        {
+            var thisField = stateMachine.GetType().GetField(StateMachineThisFieldName, BindingFlags.Instance | BindingFlags.Public);
+            return thisField?.GetValue(stateMachine);
```

Comment:
If the enclosing type (the original 'this') is generic? Will `GetValue` break? We should have tests around that (open/close generics)